GPII-2179: Pin to a published version of infusion that includes deletion relaying and new transforms.

the-t-in-rtf opened the pull request:
See [https://issues.gpii.net/browse/GPII-2179](https://issues.gpii.net/browse/GPII-2179) for details.

the-t-in-rtf committed:
NOJIRA:  Updating version in advance of next release, to ensure that no further 1.0.0 dev releases are cut.

the-t-in-rtf committed:
GPII-1815: Updated to standard common Vagrantfile and migrated `.qi.yml`.

the-t-in-rtf committed:
NOJIRA:  Updated test command to use /var/tmp for storage.

the-t-in-rtf committed:
NOJIRA: Updated browser test launch to avoid running phantomjs on Travis CI.

the-t-in-rtf committed:
NOJIRA:  Updated Travis config to test using Node 6.x and yarn.

the-t-in-rtf committed:
GPII-2179: Pin to a published version of infusion that includes the post-2.0.0 binder transform functions and deletion relay fix.

amb26 commented on .qi.yml:
```diff
@@ -1,20 +1,18 @@
+apps:
+  - app_name: gpii-pouchdb
+    git_repository: https://github.com/the-t-in-rtf/gpii-pouchdb.git
+    git_branch: GPII-1815
```
This branch reference here is a bit disturbing - what does it refer to?

the-t-in-rtf replied:
It's an artifact of the current QI system, it has to know which branch contains the QI configuration file.  As PRs with the config are merged, my assumption is that this would either be replaced by "master" or removed entirely.

amb26 replied:
This line is even more worrying now it no longer even refers to the correct branch. Shall we just delete this file entirely for now?

the-t-in-rtf committed:
KETTLE-51:  Upgrading kettle to pick up utf8 fixes.

the-t-in-rtf commented:
@amb26, this should be reviewed once the infusion and docs PRs are merged.

the-t-in-rtf committed:
NOJIRA:  Updated gpii-express to released version.  Updated outdated dependencies based on output of `yarn outdated`.

the-t-in-rtf commented:
@amb26, I propose reviewing the upcoming changes for [GPII-2267](https://issues.gpii.net/browse/GPII-2267) (improved cleanup) in this PR rather than creating a new one.

amb26 commented:
@the-t-in-rtf - sounds good to me!

the-t-in-rtf committed:
GPII-2267: All components that create directories now make a note of it and clean them up on shutdown.

the-t-in-rtf committed:
GPII-2267: Removed unused `mkdirp` dependency.

the-t-in-rtf commented:
@amb26, I have committed the changes.  In addition to confirming everything working with the existing persistence tests, I also closely monitored the temp directory during multiple runs.  All created temp directories are cleaned up.

amb26 commented on package.json:
```diff
@@ -1,36 +1,35 @@
-    "gpii-express": "1.0.0",
-    "infusion": "2.0.0-dev.20161027T155641Z.0bcd507",
+    "gpii-express": "1.0.2",
+    "infusion": "2.0.0",
```
May as well update to the infusion which is seen in your/our other current projects, which is right now 3.0.0-dev.20170131T153243Z.6aab53a

the-t-in-rtf replied:
Done, and updated kettle as well.

amb26 commented on src/js/pouch-express.js:
```diff
@@ -200,22 +205,36 @@ gpii.pouch.express.cleanup = function (that) {
-            gpii.pouch.express.initExpressPouchdb(that);
-
-            gpii.pouch.express.initDbs(that).then(function () {
+            if (that.baseDirBelongsToUs) {
```
YOU HAVE NO CHANCE TO SURVIVE MAKE YOUR TIME

colinbdclark replied:
Typo here—variable should be <code>baseDirBelongToUs</code>.

simonbates replied:
Only one base, where are the rest?

simonbates replied:
FOR GREAT JUSTICE

the-t-in-rtf replied:
Yes, thank you all, I'm glad you enjoyed it.  I decided to err on the side of also making it a meaningful and grammatically sound variable name, hence singular on "base" and plural on "belongs".  Even if you don't get the joke, it actually describes how the variable is used.

amb26 commented on src/js/pouchdb-node.js:
```diff
@@ -28,19 +27,45 @@ fluid.registerNamespace("gpii.pouch.node");
+ */
+gpii.pouch.node.cleanup = function (that) {
+    if (that.baseDirBelongsToUs) {
+        fs.rmdirSync(that.options.baseDir);
```
I'd like to see this enclosed in a no-op try/catch block to ensure that a failure doesn't interfere with any other cleanup that is on the stack - to me, this is part of the same discussion as http://stackoverflow.com/questions/130117/throwing-exceptions-out-of-a-destructor

the-t-in-rtf replied:
Done.

the-t-in-rtf committed:
GPII-2179: Updated to latest dev version of infusion and updated kettle dep.

the-t-in-rtf committed:
GPII-2179: Added “noop” try/catch so that cleanup errors can’t interfere with the rest of the component’s teardown.

amb26 commented:
Running tests/js/all-tests.js on this branch still leaves behind one PouchDB database in Temp, with the following contents:

```
C:\Users\Bosmon\AppData\Local\Temp>ls -ltr gpii-pouch-express-persistent
total 21
drwxrwx---+ 1 Bosmon None   0 Feb  3 14:50 _replicator
drwxrwx---+ 1 Bosmon None   0 Feb  3 14:50 pouch__all_dbs__
drwxrwx---+ 1 Bosmon None   0 Feb  3 14:50 _users
-rwxrwx---+ 1 Bosmon None 153 Feb  3 14:50 config.json
drwxrwx---+ 1 Bosmon None   0 Feb  3 14:50 persistence
drwxrwx---+ 1 Bosmon None   0 Feb  3 14:50 persistence-mrview-4571caff952754210024696e3967ac5b
```

Also - can you tell me how I can prevent testem from firing up Internet Explorer on my machine? I tried adding strings of various forms to this line in package.json but none of them worked:

```
"test:browser": "node node_modules/testem/testem.js ci --skip safari,phantomjs"
```

the-t-in-rtf commented:
That content is clearly from the persistence tests, which are provisioned once and reused over a couple of restarts.  I'll have to add a final bit of cleanup.

I'll also fire up a Windows VM and get back to you on the testem command.

the-t-in-rtf committed:
GPII-2179: Hardened express pouchdb cleanup.  Improved persistence tests to ensure cleanup after final run.

the-t-in-rtf commented:
I have hardened the persistence tests so that no content remains after the run.

I also confirmed that the following skips the IE tests on the standard Vagrant Windows 7 VM:

`node node_modules/testem/testem.js ci --skip IE`

The allowed values for the `skip` option can be determined using the following command:

`node node_modules/testem/testem.js launchers`

the-t-in-rtf commented:
I also double-checked that there was no problem using "IE" in combination with other skips, the following (for example) worked for me:

`node node_modules/testem/testem.js ci --skip safari,phantomjs,IE`

@amb26, this should be ready for further review.

amb26 commented:
Thanks for the help on IE skipping, those options work for me! The cleanup code is looking promising, but it looks like there is some form of race condition during the cleanup - I'm unable to run the tests to conclusion, they always finish with a hang, and then a deferred rejection failure for one or other of the files in gpii-pouch-express-persistent - here are two samples:

```
14:14:37.128:  jq: Test concluded - Module "Testing persistence within a single restart..." Test name "Confirm that we have indexed records after a restart...":
 2/2 passed - PASS
14:14:37.149:  express pouchdb instance '920mr2ob-3087' initalizing...
14:14:37.174:  express baseDir: 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent'...
14:14:37.177:  Express stopped...
14:14:37.177:  Express server listening on port 6798
14:14:37.178:  Express started...
14:14:37.181:  This database already has one or more records, no data will be loaded...
14:14:37.181:  Data loaded for database 'persistence'...
14:14:42.187:  Test case listener has not responded after 5000ms - at sequence pos 4 of 4 sequence element {
    "event": "{getAfterResetRequest}.events.onComplete",
    "listener": "jqUnit.assertEquals",
    "args": [
        "There should no longer be a record.",
        404,
        "{getAfterResetRequest}.nativeResponse.statusCode"
    ]
} of fixture Confirm that the record is no longer there after cleaning up the data and recreating express-pouchdb...
14:14:49.945:  Failure in task fixture {
    "task": "{harness}.cleanup",
    "resolve": "{getAfterResetRequest}.send",
    "resolveArgs": []
}: promise has received disposition "reject" with payload {
    "errno": -4082,
    "code": "EBUSY",
    "syscall": "unlink",
    "path": "C:\\Users\\Bosmon\\AppData\\Local\\Temp\\gpii-pouch-express-persistent\\pouch__all_dbs__\\LOG"
} when "resolve" was expected
14:14:49.949:  jq: FAIL: Module "Testing persistence within a single restart..." Test name "Confirm that the record is no longer there after cleaning up the dat
a and recreating express-pouchdb..." - Message: Failure in task fixture {
  "task": "{harness}.cleanup",
  "resolve": "{getAfterResetRequest}.send",
  "resolveArgs": []
}: promise has received disposition "reject" with payload {
  "errno": -4082,
  "code": "EBUSY",
  "syscall": "unlink",
  "path": "C:\\Users\\Bosmon\\AppData\\Local\\Temp\\gpii-pouch-express-persistent\\pouch__all_dbs__\\LOG"
} when "resolve" was expected
14:14:49.951:  jq: Source:     at pok (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.fail (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at Array.<anonymous> (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:273:20)
    at Object.that.complete (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\FluidPromises.js:70:25)
    at Object.that.reject (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\FluidPromises.js:61:22)
    at E:\source\gits\gpii-pouchdb\src\js\pouch-express.js:9:7741
    at next (E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:74:7)
    at CB (E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:110:9)
    at E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:250:18
    at next (E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:74:7)
    at CB (E:\source\gits\gpii-pouchdb\node_modules\rim .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
```

and

```
14:15:48.251:  jq: Test concluded - Module "Testing persistence within a single restart..." Test name "Confirm that we have indexed records after a restart...":
 2/2 passed - PASS
14:15:48.283:  express pouchdb instance '15zga6yo-3087' initalizing...
14:15:48.307:  express baseDir: 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent'...
14:15:48.309:  Express stopped...
14:15:48.309:  Express server listening on port 6798
14:15:48.310:  Express started...
14:15:48.312:  This database already has one or more records, no data will be loaded...
14:15:48.312:  Data loaded for database 'persistence'...
14:15:53.317:  Test case listener has not responded after 5000ms - at sequence pos 4 of 4 sequence element {
    "event": "{getAfterResetRequest}.events.onComplete",
    "listener": "jqUnit.assertEquals",
    "args": [
        "There should no longer be a record.",
        404,
        "{getAfterResetRequest}.nativeResponse.statusCode"
    ]
} of fixture Confirm that the record is no longer there after cleaning up the data and recreating express-pouchdb...
14:16:01.052:  Failure in task fixture {
    "task": "{harness}.cleanup",
    "resolve": "{getAfterResetRequest}.send",
    "resolveArgs": []
}: promise has received disposition "reject" with payload {
    "errno": -4082,
    "code": "EBUSY",
    "syscall": "unlink",
    "path": "C:\\Users\\Bosmon\\AppData\\Local\\Temp\\gpii-pouch-express-persistent\\pouch__all_dbs__\\000004.log"
} when "resolve" was expected
14:16:01.056:  jq: FAIL: Module "Testing persistence within a single restart..." Test name "Confirm that the record is no longer there after cleaning up the dat
a and recreating express-pouchdb..." - Message: Failure in task fixture {
  "task": "{harness}.cleanup",
  "resolve": "{getAfterResetRequest}.send",
  "resolveArgs": []
}: promise has received disposition "reject" with payload {
  "errno": -4082,
  "code": "EBUSY",
  "syscall": "unlink",
  "path": "C:\\Users\\Bosmon\\AppData\\Local\\Temp\\gpii-pouch-express-persistent\\pouch__all_dbs__\\000004.log"
} when "resolve" was expected
14:16:01.059:  jq: Source:     at pok (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:112:15)
    at Object.fail (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\jqUnit\js\jqUnit.js:129:13)
    at Array.<anonymous> (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:273:20)
    at Object.that.complete (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\FluidPromises.js:70:25)
    at Object.that.reject (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\FluidPromises.js:61:22)
    at E:\source\gits\gpii-pouchdb\src\js\pouch-express.js:9:7741
    at next (E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:74:7)
    at CB (E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:110:9)
    at E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:250:18
    at next (E:\source\gits\gpii-pouchdb\node_modules\rimraf\rimraf.js:74:7)
    at CB (E:\source\gits\gpii-pouchdb\node_modules\rim .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
```

System is Windows 7 x64, node 6.9.2 64-bit

the-t-in-rtf commented:
Thanks, I will test with the standard windows VM to confirm, but I suspect I know exactly what's happening, as I had to migrate away from using fs.rmdir in another context for what I suspect are similar reasons.

the-t-in-rtf committed:
GPII-2179: Hardened directory removal in node pouchdb grade.

the-t-in-rtf committed:
GPII-2179:  Increased `maxBusyTries` for directory deletion and added definitions for dbs used by express-pouch.

the-t-in-rtf committed:
GPII-2179: reduced `maxBusyTries`.

the-t-in-rtf committed:
GPII-2179: Removed `rimraf` opts.

amb26 commented:
Some improvement, but I now get either the following hang:

```
15:05:42.849:  Test case listener has not responded after 5000ms - at sequence pos 4 of 4 sequence element {
    "event": "{getAfterResetRequest}.events.onComplete",
    "listener": "jqUnit.assertEquals",
    "args": [
        "There should no longer be a record.",
        404,
        "{getAfterResetRequest}.nativeResponse.statusCode"
    ]
} of fixture Confirm that the record is no longer there after cleaning up the data and recreating express-pouchdb...
```

or else the following I/O error:

```
15:04:38.432:  This database already has one or more records, no data will be loaded...
15:04:38.432:  Data loaded for database 'persistence'...
Unhandled rejection OpenError: IO error: C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent\_users/MANIFEST-000001: The system cannot find the fil
e specified.

    at E:\source\gits\gpii-pouchdb\node_modules\levelup\lib\levelup.js:119:34
    at E:\source\gits\gpii-pouchdb\node_modules\leveldown\node_modules\abstract-leveldown\abstract-leveldown.js:39:16

15:04:38.795:  Removed temporary directory 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent'...
15:04:38.797:  Sending a GET request to: /persistence/new on port 6798
15:04:38.818:  jq: Test concluded - Module "Testing persistence within a single restart..." Test name "Confirm that the record is no longer there after cleaning
 up the data and recreating express-pouchdb...": 1/1 passed - PASS
15:04:38.840:  express pouchdb instance '21anfpy6-3722' initalizing...
15:04:38.841:  Creating directory 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent' for express pouchdb instance '21anfpy6-3722'...
15:04:38.845:  jq: FAIL: Module "Testing persistence within a single restart..." Test name "Confirm that we have no indexed records after a reset..." - Message:
 Died on test #1     at Object.asyncTest (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\lib\qunit\js\qunit.js:405:9)
    at E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:651:29
    at Object.fluid.each (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:513:17)
    at Object.fluid.test.processTestCase (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:625:15)
    at E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:674:24
    at Object.fluid.each (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:513:17)
    at Object.fluid.test.processTestCaseHolder (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:667:15)
    at E: .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
15:04:38.848:  jq: Source: Error: EPERM: operation not permitted, mkdir 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent'
    at Error (native)
    at Object.fs.mkdirSync (fs.js:922:18)
    at gpii.pouch.express.initExpressPouchdb (E:\source\gits\gpii-pouchdb\src\js\pouch-express.js:9:2502)
    at togo (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\FluidIoC.js:1796:33)
    at fire (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1496:40)
    at Object.fire (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1496:40)
    at Object.fluid.initComponent (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:2675:30)
    at creator (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1929:30)
    at Object.fluid.invokeGlobalFunction (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1113:25)
    at Object.fluid.initSubcomponentImpl (E:\source\ .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
15:04:38.850:  Express stopped...
15:04:38.864:  jq: Test concluded - Module "Testing persistence within a single restart..." Test name "Confirm that we have no indexed records after a reset..."
: 0/1 passed - FAIL
15:04:38.883:  express pouchdb instance '21anfpy6-3767' initalizing...
15:04:38.883:  Creating directory 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent' for express pouchdb instance '21anfpy6-3767'...
15:04:38.887:  jq: FAIL: Module "Testing persistence within a single restart..." Test name "Clean up at the end of the run..." - Message: Died on test #1     at
 Object.asyncTest (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\lib\qunit\js\qunit.js:405:9)
    at E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:651:29
    at Object.fluid.each (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:513:17)
    at Object.fluid.test.processTestCase (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:625:15)
    at E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:674:24
    at Object.fluid.each (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:513:17)
    at Object.fluid.test.processTestCaseHolder (E:\source\gits\gpii-pouchdb\node_modules\infusion\tests\test-core\utils\js\IoCTestUtils.js:667:15)
    at E:\source\gits\gpii-pouch .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
15:04:38.890:  jq: Source: Error: EPERM: operation not permitted, mkdir 'C:\Users\Bosmon\AppData\Local\Temp\gpii-pouch-express-persistent'
    at Error (native)
    at Object.fs.mkdirSync (fs.js:922:18)
    at gpii.pouch.express.initExpressPouchdb (E:\source\gits\gpii-pouchdb\src\js\pouch-express.js:9:2502)
    at togo (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\FluidIoC.js:1796:33)
    at fire (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1496:40)
    at Object.fire (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1496:40)
    at Object.fluid.initComponent (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:2675:30)
    at creator (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1929:30)
    at Object.fluid.invokeGlobalFunction (E:\source\gits\gpii-pouchdb\node_modules\infusion\src\framework\core\js\Fluid.js:1113:25)
    at Object.fluid.initSubcomponentImpl (E:\source\ .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
15:04:38.906:  jq: Test concluded - Module "Testing persistence within a single restart..." Test name "Clean up at the end of the run...": 0/1 passed - FAIL
15:04:43.822:  Test case listener has not responded after 5000ms - at sequence pos 2 of 5 sequence element {
    "listener": "fluid.identity",
    "event": "{testEnvironment}.events.onFixturesConstructed"
} of fixture Confirm that we have no indexed records after a reset...
15:04:43.866:  Test case listener has not responded after 5000ms - at sequence pos 2 of 3 sequence element {
    "listener": "fluid.identity",
    "event": "{testEnvironment}.events.onFixturesConstructed"
} of fixture Clean up at the end of the run...
```

the-t-in-rtf committed:
GPII-2179:  Reduced express-pouchdb cleanup error to warning (now corresponds to the handling in the node pouchdb grade).

the-t-in-rtf committed:
GPII-2179: Increased `hangWait` for persistence tests to allow time for directory removal.

the-t-in-rtf committed:
GPII-2179: Removed “belt and braces” db definitions in favor of express-pouchdb-level cleanup.

the-t-in-rtf committed:
GPII-2179:  Added braces to go with my belt.

the-t-in-rtf committed:
GPII-2179: Ensured that only one db cleanup pass at a time executes.

amb26 commented:
Tests now pass again, but sadly we are back to the base directory not being cleaned up:
```
15:19:25.071:  Data loaded for database 'pouch__all_dbs__'...
15:19:37.968:  ERROR: Unable to remove base directory...
{
    "errno": -4082,
    "code": "EBUSY",
    "syscall": "unlink",
    "path": "C:\\Users\\Bosmon\\AppData\\Local\\Temp\\gpii-pouch-express-persistent\\_users\\000003.log"
}
```

and

```
15:19:51.546:  ERROR: Unable to remove base directory...
{
    "errno": -4082,
    "code": "EBUSY",
    "syscall": "unlink",
    "path": "C:\\Users\\Bosmon\\AppData\\Local\\Temp\\gpii-pouch-express-persistent\\_users\\LOCK"
}
15:19:51.575:  Express stopped...
```

I imagine there's a fault in gpii-pouch-express whereby it never actually closes the file handle?

the-t-in-rtf committed:
GPII-2179: Found the origin of the race condition causing test warnings on Windows.

the-t-in-rtf committed:
GPII-2179:  Added timeout to hopefully allow the _users db to finish shutdown before we perform the main cleanup.

the-t-in-rtf committed:
GPII-2179: Improved logging to help track down windows warnings further.

the-t-in-rtf committed:
GPII-2179:  Additional tweaks to try to get windows persistence tests to clean up after themselves.

the-t-in-rtf committed:
GPII-2179:  Removed two previous ineffective cleanup approaches.  Added a new one.

the-t-in-rtf committed:
GPII-2179:  Continued down the “remove unused parts of express-pouchdb” rabbit hole.

the-t-in-rtf commented:
As @amb26 pointed out, there are problems with the content in the _users database not being removed when the persistence tests are run.  After long investigation, I don't have an explanation yet, but I can confirm that it's only a problem with the persistence tests.  My hope is that by looking at the options and standard cleanup sequence steps used in the rest of the tests, we can hopefully find some clue.

the-t-in-rtf committed:
GPII-2179: Tried using events rather than promises for persistence cleanup.

amb26 commented:
I'd be fine with a "fake test fixture" that runs last and just blows away the surplus directory, plus a writeup of the problem

the-t-in-rtf committed:
GPII-2179:  Added additional cleanup step to hopefully clean up straggling windows content.

the-t-in-rtf committed:
GPII-2179: Tried alternate listener-based cleanup (and adding logging).

the-t-in-rtf committed:
GPII-2179: Tweaked logging further.

the-t-in-rtf committed:
GPII-2179: Moar logs.

the-t-in-rtf committed:
GPII-2179: Trying to work around ongoing errors with cleanup.

the-t-in-rtf committed:
GPII-2179:  Moved final cleanup to `posttest` steps.

the-t-in-rtf commented:
@amb26, after discovering quite a few ways to not actually remove anything, I ended up adding a "posttest" step.  This will handle all test runs that use "npm test", including CI builds.  I hope this will be adequate.

amb26 commented:
Thanks - the gpii-persistent dir is now cleaned up correctly. However, I do find what looks like 2 further directories created per run, I assume by testem - their names are of the form testem-nnnn and scoped_dir_nnnn_mmmm. I've linked to a gist containing some here:

https://gist.github.com/amb26/f5a648358153fb506448b451f603ac64
https://gist.github.com/amb26/0d5b63a26cac0a6ea4a0e3d8be09036a

Again I'd be happy with a blunt instrument that just makes best efforts to remove any dirs in temp whose names match this pattern in your posttest.

the-t-in-rtf commented:
@amb26, I looked at the testem docs and then the code, starting with the [local function provided by the testem prototype](https://github.com/testem/testem/blob/241b1e42e96cd7e4d656efda6475e96e1f7978ae/lib/config.js#L425).  I searched the whole project for direct and indirect uses of that user directory, it looks like they don't provide a built-in option or code to handle this bit of cleanup.  Looking at their "launcher", I can see that the default is basically os.tmpdir plus the "testem-" prefix and a unique id, so a simple pattern match should be adequate.

I have already been seriously thinking about maturing my previous work with testem into a micromodule, as I believe it would help drive discussions around code coverage within testem, and would give us a new option for using QUnit browser tests in combination with node-side fixtures.

Rather than solving this per project, it makes sense to provide a cleanup mechanism as part of that package.

For the cleanup, I'd create a static function that can clean up after testem as part of that package.  That function would be wired into the shutdown cycle of [the testem component of the kind you've seen in my work on gpii-binder](https://github.com/the-t-in-rtf/gpii-binder/blob/GPII-2077/testem.js).  We also need a painless way to trigger the cleanup mechanism from testem.json configuration files.  My idea is to use a "bin" definition in package.json and provide docs on calling that from testem.json.

I'll go ahead and write up my thoughts on the "gpii-testem" micromodule and share that ticket here.  If you're happy with that as the long term solution, I'll write that up and use it (or at least its cleanup) with everything I maintain that uses testem, including this package.

the-t-in-rtf commented:
@amb26, I [posted some thoughts to the mailing list](http://lists.gpii.net/pipermail/architecture/2017-February/004500.html) and [wrote up a ticket with my proposed approach](https://issues.gpii.net/browse/GPII-2296).

the-t-in-rtf committed:
GPII-2296: Updated to use gpii-testem, and to compile cross-environment coverage reports.

the-t-in-rtf commented:
@amb26, as discussed, I have updated the pull to use gpii-testem.  Running `npm test` will now result in cross-environment coverage reporting.

the-t-in-rtf commented:
This also addresses the Testem cleanup concerns.

the-t-in-rtf committed:
NOJIRA:  Fixed goofy “bad bullets” in README.

amb26 commented:
The testem tests hang in Chrome at the end of the run - after reaching test 13/13, the browser window stays open and the test run does not proceed any further. If I skip Chrome, IE shows the same visible behaviour but in addition reports this disconnection:

ok 5 IE 11.0 - Common tests for `gpii.pouch` component...: Test the `POST` method (without an ID)...
not ok 6 IE - error
    ---
        message: >
            Error: Browser exited unexpectedly

        Log: |
            { type: 'error', text: 'Error: Browser exited unexpectedly' }
    ...

1..6
# tests 6
# pass  5
# skip  0
# fail  1

The exit is reported at a randomly shifting position each run.

Running just the node tests shows that the cleanup is now nearly perfect - however there is now just one directory left behind in Temp, whose name is reports-1caovdlk-31 , containing a report.tap file

the-t-in-rtf committed:
GPII-2179:  Updated dependencies, including gpii-testem.

the-t-in-rtf committed:
GPII-2179: “Pinned” updated dependencies.

the-t-in-rtf committed:
GPII-2179:  Migrated to using git:// URL for forked express-pouchdb dependency, to avoid trust issues with GitHub.

the-t-in-rtf committed:
NOJIRA: Updated secondary forked dependency to use git:// rather than https://

the-t-in-rtf committed:
GPII-2179: Updated to newer version of gpii-testem to pick up released version of gpii-express throughout.

the-t-in-rtf committed:
GPII-2179:  Updated `gpii-testem` dependency to avoid problems caused by the lack of `mkdirp`.

the-t-in-rtf committed:
Moved reports dir to local directory.

the-t-in-rtf commented:
The report.tap file is not testing the ability to test, but actually represents machine-readable output from the test run itself.  It should therefore be preserved.  I have updated to leave it in "reports", as we've done in other projects.

The other failure seems to be a failure in calling the Testem callback at the end of a browser run.  On a hunch, I tried it in my Bootcamp VM (not a QI VM), and see the same hangs there.  Will investigate further, although it might be good to cover this in our next pairing session.

the-t-in-rtf committed:
GPII-2179: Updated to newer version of `gpii-testem` to hopefully address Windows test stability issues.

the-t-in-rtf commented:
Now that the gpii-testem errors are resolved, I am seeing a test failure in IE, which seems to be down to a difference in the way the PouchDB "info" method works.  Troubleshooting now, I will advise when this is ready for review again.

the-t-in-rtf commented:
I was seeing intermittent failures in the IE tests, particularly around the "info" method.  I only saw errors when running all tests at once.  By the time I finished working my way through running just the IE tests (which passed, repeatedly), then just the browser tests (which passed, repeatedly), the overall tests worked, I haven't observed a failure in 5 consecutive runs.

@amb26, are the tests running on your machine?  If so, I'd propose creating a ticket with the symptoms of the intermittent failures and tracking that if it comes up more frequently.

amb26 commented:
The Chrome tests now run and exit properly, but I get a browser disconnect from IE at some (early) point during its tests, as follows:

```
14:47:44.571:  ""
ok 1 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `info` method...
ok 2 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `POST` method (with an ID)...
ok 3 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `GET` method (with a non-existent ID)...
ok 4 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `PUT` method (with an ID)...
ok 5 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `POST` method (without an ID)...
ok 6 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `bulkDocs` function...
ok 7 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `allDocs` function...
ok 8 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `bulkGet` function...
ok 9 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `remove` function...
ok 10 Chrome 56.0 - Common tests for `gpii.pouch` component...: Test the `remove` method (with a non-existent ID)...
ok 11 Chrome 56.0 - Common tests for `gpii.pouch` component...: Testing `query` handling (passing a single map function)...
ok 12 Chrome 56.0 - Common tests for `gpii.pouch` component...: Testing `query` handling (passing a map function as part of a JSON document)...
ok 13 Chrome 56.0 - Common tests for `gpii.pouch` component...: Testing `query` handling (passing both map and reduce function)...
ok 14 IE 11.0 - Common tests for `gpii.pouch` component...: Test the `info` method...
ok 15 IE 11.0 - Common tests for `gpii.pouch` component...: Test the `POST` method (with an ID)...
ok 16 IE 11.0 - Common tests for `gpii.pouch` component...: Test the `GET` method (with a non-existent ID)...
not ok 17 IE - error
    ---
        message: >
            Error: Browser exited unexpectedly

        Log: |
            { type: 'error', text: 'Error: Browser exited unexpectedly' }
    ...

1..17
# tests 17
# pass  16
# skip  0
# fail  1
```

Windows 7 x64, IE version 11.0.9600.17843

the-t-in-rtf commented:
The [Windows 7 VM](https://atlas.hashicorp.com/inclusivedesign/boxes/windows7-eval-x64) has IE 11.0.9600.18537.  Unfortunately I am struggling to get pouchdb and leveldown to build.  @amb26, what (if anything) did you have to do before you could run "npm install" in this project?

the-t-in-rtf committed:
GPII-2179: Updated to latest gpii-testem and moved testem.js to avoid potentially clobberring the client-side /testem.js file.

the-t-in-rtf commented:
OK, I am having problems running the node tests in a VM (I can't find a package that will install a usable vcbuild.exe).  I noticed some initial behavior that suggested gpii-testem did not include recent improvements. 
So, I updated gpii-testem to the latest dev release, and after that I am able to run the browser tests with Windows 7 and IE 11.0.9600.18537.   @amb26, please update and confirm if the browser tests work for you now.

the-t-in-rtf committed:
GPII-2380: Update to "root safe dedupe" version of infusion.  NOJIRA:  Resolve remaining Vagrant build issues.

the-t-in-rtf commented:
In pushing through further changes ("root safe dedupe infusion"), I worked further with the Vagrant config.  The tests now pass on Windows 10, including in IE, with the following caveat:

Until [https://issues.gpii.net/browse/GPII-2341](https://issues.gpii.net/browse/GPII-2341) is resolved, the process for running the tests in Vagrant is as follows:

1. Run "vagrant up" from the host machine.
2. When this process completes, open the VM and open Internet Explorer once.
3. Click through the dialog to set the security settings. 
4. When closing the browser, tick the box to automatically close multiple tabs.
5. Run "vagrant ci test" from the host machine

the-t-in-rtf commented:
In the Vagrant Windows 10 VM, the tests are confirmed passing with IE 11.953.14393.0

the-t-in-rtf committed:
NOJIRA: Removed outdated QI config.

the-t-in-rtf committed:
GPII-2380: Updated to "root safe dedupe" versions of infusion and kettle, including in dependencies.

the-t-in-rtf committed:
GPII-2392: Added timeouts and graceful-fs wrapper to rimraf calls.

the-t-in-rtf committed:
GPII-2380: Updated to kettle 1.4.1 and "root safe dedupe" version of gpii-express.
NOJIRA: Updated additional dependencies prior to review and release. Pinned mock-fs dependency.

the-t-in-rtf committed:
GPII-2392: Converted secondary cleanup error to warning.

the-t-in-rtf committed:
GPII-2392: Tidied up timeout warning...

the-t-in-rtf committed:
NOJIRA: Updated Travis config to hopefully fix leveldown install failures.

the-t-in-rtf committed:
Added initial GitLab CI configuration

amb26 commented:
The tests run well, but the following junk was left behind in Temp:

```
drwx------+ 1 Bosmon None   4096 Apr 24 15:10 qo0ul6m-184
drwx------+ 1 Bosmon None   4096 Apr 24 15:10 qo0ul6m-264
drwx------+ 1 Bosmon None   4096 Apr 24 15:10 qo0ul6m-339
-rwx------+ 1 Bosmon None   2578 Apr 24 15:11 log.txt
drwx------+ 1 Bosmon None      0 Apr 24 15:11 user_data_dir-4vhdetmh-31
-rwx------+ 1 Bosmon None 838067 Apr 24 15:11 CF13.tmp
-rwx------+ 1 Bosmon None  25927 Apr 24 15:11 D1B4.tmp
drwx------+ 1 Bosmon None      0 Apr 24 15:11 scoped_dir_8236_29761
```

amb26 commented:
So, for the failed cleanup for the directory with suffix "-184", here is the relevant section of the logs - which make no mention of cleanup, either successful or failed:

```
15:10:57.659:  express pouchdb instance 'qo0ul6m-184' initalizing...
15:10:57.663:  Creating directory 'C:\Users\Bosmon\AppData\Local\Temp\qo0ul6m-184' for express pouchdb instance 'qo0ul6m
-184'...
15:10:57.996:  express baseDir: 'C:\Users\Bosmon\AppData\Local\Temp\qo0ul6m-184'...
15:10:58.013:  Express server listening on port 9595
15:10:58.016:  Express started...
15:10:58.185:  Data loaded for database 'rgb'...
15:10:58.205:  DataSource Issuing HTTP request with options {
    "port": "9595",
    "method": "GET",
    "headers": {

    },
    "protocol": "http:",
    "auth": null,
    "host": "localhost:9595",
    "hostname": "localhost",
    "path": "/rgb/_design/rgb/_view/byColor?key=%22red%22",
    "url": "http://localhost:9595/rgb/_design/rgb/_view/byColor",
    "termMap": {

    },
    "directModel": {
        "key": "red"
    },
    "operation": "get",
    "reverse": false,
    "writeMethod": "PUT"
}
15:10:58.574:  jq: Test concluded - Module "Integration tests for gpii-pouch and the 'url encoding' express dataSource g
rade..." Test name "We should be able to retrieve a record using a single key...": 3/3 passed - PASS
```

the-t-in-rtf commented:
Thanks, I will attempt to reproduce the cruft problems on my machine and then Windows VM.

the-t-in-rtf commented:
Thanks for the tip about the second half of the component id being deterministic.  The same three tests are consistently leaving cruft behind on a clean Windows 10 VM.  I will investigate further.

the-t-in-rtf commented:
As soon as I navigated to the temporary directory on my Mac, I could see that the same three ID segments (184, 264, 339) were also left behind.  Looking at the logs, I confirmed that all three were tests in ` tests/js/express-dataSource-integration-tests.js`.  Those tests were based off of a different caseHolder than the rest, and that specific caseHolder lacked the usual cleanup steps.  I updated the caseHolder and confirmed working on OS X.  Windows is another story.  There, a reasonable 1-2 second timeout results in failures to clean up those directories.  Setting the timeout to 60 seconds reveals a rimraf error about 15 seconds in:

```
Error removing temporary directory: { Error: EBUSY: resource busy or locked, unlink 'C:\Users\vagrant\AppData\Local\Temp\aq988kpz-184\_replicator\000003.log'
==> windows10:     at Error (native)
==> windows10:   errno: -4082,
==> windows10:   code: 'EBUSY',
==> windows10:   syscall: 'unlink',
==> windows10:   path: 'C:\\Users\\vagrant\\AppData\\Local\\Temp\\aq988kpz-184\\_replicator\\000003.log' }
```
I will review and see what I can do about that, but just to confirm, is a clearly logged timeout acceptable?  Is a 15 second timeout acceptable?  Either would be fairly easy.

the-t-in-rtf committed:
GPII-2179:  Updated dataSource integration tests to clean up after themselves.

the-t-in-rtf commented:
OK, it turns out that this was entirely a byproduct of the dataSource integration tests having a separate caseHolder.  The caseHolder used in most tests in this package:

1. Cleans up after itself after each run.
2. Instantiates the databases that express-pouchdb uses so that it can clean them up afterward.

After starting down a few longer roads, I realized it was indeed just that simple, applied the fix, and confirmed that the content is cleared out on both OS X and Windows 10.

@amb26, this should be ready for further review.

amb26 commented:
Thanks, the situation is hugely improved with these fixes and I think this is now mergeable. For future reference, the files left behind in Temp on my system are now the following 2:

```
-rwxrwx---+ 1 Bosmon         None  2578 Apr 26 15:26 log.txt
drwxrwx---+ 1 Bosmon         None     0 Apr 26 15:26 user_data_dir-arxs6l3d-31
```

I imagine the latter is a pouch-express artefact, and I notice that I do now receive useful diagnostics from failed rimraf runs:

```
Error removing temporary directory: Rimraf failed to complete within 1000 milliseconds...
```
In fact I seem to receive more of these diagnostics than might be justified, given that there is only one directory for which cleanup has visibly failed.

the-t-in-rtf commented:
Thanks for the merge.  There is no attempt to prevent rimraf from completing its work even if the timeout fires (we just prevent it from resolving/rejecting the promise if it already has a disposition).  I suspect that rimraf is managing to finish up in those cases in the time between the timeout and the teardown of the test fixtures.

The user_data_dir is in fact gpii-testem's fault, I have created a separate ticket to track that work: [GPII-2396](https://issues.gpii.net/browse/GPII-2396).

The log.txt file is indeed an omission in the express-pouchdb component cleanup, I have created a ticket here for that work: [GPII-2397}(https://issues.gpii.net/browse/GPII-2397)
